src/mapogcsld.cpp

```diff
@@ -196,6 +249,206 @@ static void msSLDApplySLD_DuplicateLayers(mapObj *map, int nSLDLayers,
+
+    lp->labelitem = msStrdup(sldLayer->labelitem);
+  }
+
+  if (sldLayer->classitem) {
+    if (lp->classitem)
+      free(lp->classitem);
+
+    lp->classitem = msStrdup(sldLayer->classitem);
+  }
+
+  /* opacity for sld raster */
+  if (lp->type == MS_LAYER_RASTER && sldLayer->compositer &&
+      sldLayer->compositer->opacity != 100)
+    msSetLayerOpacity(lp, sldLayer->compositer->opacity);
+
+  /* mark as auto-generate SLD */
+  if (lp->connectiontype == MS_WMS)
+    msInsertHashTable(&(lp->metadata), "wms_sld_body", "auto");
+
+  /* The SLD might have a FeatureTypeConstraint */
+  if (sldLayer->filter.type == MS_EXPRESSION) {
+    if (lp->filter.string && lp->filter.type == MS_EXPRESSION) {
+      char *pszBuffer = msStringConcatenate(NULL, "((");
+      pszBuffer = msStringConcatenate(pszBuffer, lp->filter.string);
+      pszBuffer = msStringConcatenate(pszBuffer, ") AND (");
+      pszBuffer = msStringConcatenate(pszBuffer, sldLayer->filter.string);
+      pszBuffer = msStringConcatenate(pszBuffer, "))");
+      msFreeExpression(&lp->filter);
+      msInitExpression(&lp->filter);
+      lp->filter.string = pszBuffer;
+      lp->filter.type = MS_EXPRESSION;
+    } else {
+      msFreeExpression(&lp->filter);
+      msInitExpression(&lp->filter);
+      lp->filter.string = msStrdup(sldLayer->filter.string);
+      lp->filter.type = MS_EXPRESSION;
+    }
+  }
+
+  /*in some cases it would make sense to concatenate all the class
+    expressions and use it to set the filter on the layer. This
+    could increase performance. Will do it for db types layers
+    #2840*/
+  if (lp->filter.string == NULL ||
+      (lp->filter.string && lp->filter.type == MS_STRING && !lp->filteritem)) {
+    if (lp->connectiontype == MS_POSTGIS ||
+        lp->connectiontype == MS_ORACLESPATIAL ||
+        lp->connectiontype == MS_PLUGIN) {
+      if (lp->numclasses > 0) {
+        /* check first that all classes have an expression type.
+          That is the only way we can concatenate them and set the
+          filter expression */
+        int k;
+        for (k = 0; k < lp->numclasses; k++) {
+          if (lp->_class[k]->expression.type != MS_EXPRESSION)
+            break;
+        }
+        if (k == lp->numclasses) {
+          char szTmp[512];
+          char *pszBuffer = NULL;
+          for (k = 0; k < lp->numclasses; k++) {
+            if (pszBuffer == NULL)
+              snprintf(szTmp, sizeof(szTmp), "%s",
+                       "(("); /* we a building a string expression,
+                                 explicitly set type below */
+            else
+              snprintf(szTmp, sizeof(szTmp), "%s", " OR ");
+
+            pszBuffer = msStringConcatenate(pszBuffer, szTmp);
+            pszBuffer = msStringConcatenate(pszBuffer,
+                                            lp->_class[k]->expression.string);
+          }
+
+          snprintf(szTmp, sizeof(szTmp), "%s", "))");
+          pszBuffer = msStringConcatenate(pszBuffer, szTmp);
```

Comment:
could be simplified with std::string concatenation

Reply:
This code was from the original SLD logic, but moved to a new function. Will review.

Reply:
yes I realized that. no obligation to fix that. up to you